### PR TITLE
[#1747] feat(remote merge): Introduce common merger to merge multiple data streams.

### PR DESCRIPTION
**Pull request by zhengchenyu**
### What changes were proposed in this pull request?

Introduce merger to merge multiple data streams according to key. Minimum heap K-way merge sorting is used to merge and sort the data streams that have been partially sorted.

### Why are the changes needed?

Fix: #1747

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

unit test and test in cluster.

### CODE REVIEW

**Comment by jerqi** [comment]
Will be these methods used for  client?

**Review by jerqi — APPROVED** [review]
LGTM.

File: common/src/main/java/org/apache/uniffle/common/merger/Merger.java [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+      RecordsWriter<K, V> writer =
+          new RecordsWriter<K, V>(rssConf, output, keyClass, valueClass, raw);
+      boolean recorded = true;
+      while (this.next()) {
+        writer.append(this.getCurrentKey(), this.getCurrentValue());
+        if (output instanceof Recordable) {
+          recorded =
+              ((Recordable) output)
+                  .record(writer.getTotalBytesWritten(), () -> writer.flush(), false);
+        }
```
Comment (jerqi) [inline common/src/main/java/org/apache/uniffle/common/merger/Merger.java:176]: We should throw an exception?

File: common/src/main/java/org/apache/uniffle/common/merger/Merger.java [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    }
+
+    @Override
+    public Object getCurrentValue() {
+      return currentValue;
+    }
+
+    @Override
+    public boolean next() throws IOException {
+      if (size() == 0) {
```
Comment (jerqi) [inline common/src/main/java/org/apache/uniffle/common/merger/Merger.java:123]: isEmpty()?

Comment (zhengchenyu) [reply common/src/main/java/org/apache/uniffle/common/merger/Merger.java:123]: @jerqi The parent class is org.apache.hadoop.util.PriorityQueue, PriorityQueue does not have isEmpty method.

**Review by jerqi — APPROVED** [review]
LGTM.

**Pull request merged**